### PR TITLE
copr: Add non-vectorized json-length

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Wangweizhen <hawking.rei@gmail.com>

PCP #5751

###  What have you changed?

Add non-vectorized json-length

###  What is the type of the changes?

- New feature

###  How is the PR tested?

Unit tested. Will run copr test later.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No. It should not change the behavior.

###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test

File: components/tidb_query/src/expr/builtin_json.rs [hunk 1]
```diff
@@ -194,6 +208,41 @@ mod tests {
...
+            let json = datum_expr(match input {
+                None => Datum::Null,
+                Some(s) => Datum::Json(s.parse().unwrap()),
+            });
+            let param = datum_expr(match param {
+                None => Datum::Null,
+                Some(b) => Datum::Bytes(b),
+            });
+
+            let op = scalar_func_expr(ScalarFuncSig::JsonLengthSig, &[json, param]);
```
Comment (sticnarf) [inline components/tidb_query/src/expr/builtin_json.rs]: The test doesn't seem totally correct. It tests the case when the second argument is null (which shouldn't happen IMO), but missing the case when there is only one argument.

**Comment by breeswish** [comment]
@sticnarf PTAL

**Comment by breeswish** [comment]
@hawkingrei Could you address the comment about the case which is there is only 1 argument?

**Comment by hawkingrei** [comment]
@breeswish @sticnarf  Completed to test.  PLAT

**Comment by hawkingrei** [comment]
@breeswish  Fixed conflict.

**Comment by breeswish** [comment]
Thanks! Could you take a look about this review comment? https://github.com/tikv/tikv/pull/5916/commits/eaee2cb0fe41b3c904106cde8f114f9df9fcad17#r347750940

File: components/tidb_query/src/expr/builtin_json.rs (hunk 1 above)
Comment (hawkingrei) [reply components/tidb_query/src/expr/builtin_json.rs]: done

**Comment by hawkingrei** [comment]
I add a test that all the parameter is None. so I have fixed the issue from sticnarf’s review. But I cannot see any code review from the page of [eaee2cb#r347750940](https://github.com/tikv/tikv/commit/eaee2cb0fe41b3c904106cde8f114f9df9fcad17#r347750940).

**Comment by breeswish** [comment]
Hi, @hawkingrei   I think @sticnarf was saying that test cases are missing for insufficient number of arguments (instead of arguments in `None`).

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/105

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/105

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/105

**Comment by hawkingrei** [comment]
JSON_LENGTH have two arguments. the second argument is optional. so the first argument is Some(json_text) or None. the second argument is Some(json_path) or None,too.So an insufficient number of arguments is possible.

File: components/tidb_query/src/expr/builtin_json.rs [hunk 2]
```diff
@@ -81,6 +81,20 @@ impl ScalarFunc {
...
     }
 
+    pub fn json_length<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<i64>> {
+        let j = try_opt!(self.children[0].eval_json(ctx, row));
+        let parser = JsonFuncArgsParser::new(row);
+        let path_exprs: Vec<_> = match parser.get_path_exprs(ctx, &self.children[1..])? {
```
Comment (niedhui) [inline components/tidb_query/src/expr/builtin_json.rs]: I think when the second argument is not provided, the length of `self.children`  is 1, `self.children[1..]` will panic. 
so you need to add a test case like `let op = scalar_func_expr(ScalarFuncSig::JsonLengthSig, &[json]);`

File: components/tidb_query/src/expr/builtin_json.rs [hunk 3]
```diff
@@ -204,6 +218,115 @@ mod tests {
...
+                Some(b"$.a.a".to_vec()),
+                Some(1),
+            ),
+            (
+                Some(r#"{"a": [1, 2, {"aa": "xx"}]}"#),
+                Some(b"$.a[2].aa".to_vec()),
+                Some(1),
+            ),
+            // Tests without path expression
+            (Some(r#"{}"#), None, Some(0)),
```
Comment (niedhui) [inline components/tidb_query/src/expr/builtin_json.rs]: for this case,  the length of args is 2 and the second argument is Null, we should return null;
```
mysql> select json_length("[1,2,[1,3]]");
+----------------------------+
| json_length("[1,2,[1,3]]") |
+----------------------------+
|                          3 |
+----------------------------+
1 row in set (0.00 sec)

mysql> select json_length("[1,2,[1,3]]", NULL);
+----------------------------------+
| json_length("[1,2,[1,3]]", NULL) |
+----------------------------------+
|                             NULL |
+----------------------------------+
1 row in set (0.00 sec)
```

File: components/tidb_query/src/expr/builtin_json.rs (hunk 2 above)
Comment (hawkingrei) [reply components/tidb_query/src/expr/builtin_json.rs]: Thank you. I wrongly understand ```scalar_func_expr```.

File: components/tidb_query/src/expr/builtin_json.rs (hunk 3 above)
Comment (hawkingrei) [reply components/tidb_query/src/expr/builtin_json.rs]: Done

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/105

File: components/tidb_query/src/expr/builtin_json.rs [hunk 4]
```diff
@@ -204,6 +218,115 @@ mod tests {
     use crate::expr::{EvalContext, Expression};
     use tipb::ScalarFuncSig;
 
+    #[test]
+    fn test_json_length() {
+        let cases = vec![
+            (None, None, None),
+            (Some("null"), None, None),
```
Comment (niedhui) [inline components/tidb_query/src/expr/builtin_json.rs]: hi, here are some corner cases we need to test. and current implementation seems can not pass

```rust
// when the second argument is null, the result should always be null
(Some(r#"{}"#), Some(Datum::Null), None)
// only one argument, and is string "null", result should be 1
(Some("null"), None, Some(1)),
```

the related SQL:

```msyql
mysql> select json_length("{}", null);
+-------------------------+
| json_length("{}", null) |
+-------------------------+
|                    NULL |
+-------------------------+
1 row in set (0.00 sec)


mysql> select json_length("null");
+---------------------+
| json_length("null") |
+---------------------+
|                   1 |
+---------------------+
1 row in set (0.00 sec)

```

Comment (hawkingrei) [reply components/tidb_query/src/expr/builtin_json.rs]: Thank you. Fixed

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/105

File: components/tidb_query/src/codec/mysql/json/json_length.rs [hunk 5]
```diff
@@ -73,9 +73,9 @@ mod tests {
             (r#"{"a": [1, 2, {"aa": "xx"}]}"#, Some("$[*]"), None),
             (r#"{"a": [1, 2, {"aa": "xx"}]}"#, Some("$**.a"), None),
             // Tests path expression does not identify a section of the target document
-            (r#"{"a": [1, 2, {"aa": "xx"}]}"#, Some("$.c"), None),
-            (r#"{"a": [1, 2, {"aa": "xx"}]}"#, Some("$.a[3]"), None),
-            (r#"{"a": [1, 2, {"aa": "xx"}]}"#, Some("$.a[2].b"), None),
+            //(r#"{"a": [1, 2, {"aa": "xx"}]}"#, Some("$.c"), None),
```
Comment (niedhui) [inline components/tidb_query/src/codec/mysql/json/json_length.rs]: Forgot to uncomment these lines? 😃

File: components/tidb_query/src/expr/builtin_json.rs [hunk 6]
```diff
@@ -307,6 +308,8 @@ mod tests {
                 None,
             ),
             // Tests path expression does not identify a section of the target document
+            /*
```
Comment (niedhui) [inline components/tidb_query/src/expr/builtin_json.rs]: ditto

Comment (hawkingrei) [reply components/tidb_query/src/expr/builtin_json.rs]: Done

File: components/tidb_query/src/codec/mysql/json/json_length.rs (hunk 5 above)
Comment (hawkingrei) [reply components/tidb_query/src/codec/mysql/json/json_length.rs]: 😂Done

**Comment by hawkingrei** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/105

**Comment by hawkingrei** [comment]
@breeswish  Why doesn't the sre-bot prompt that the task is completed and get score?

**Comment by you06** [comment]
@hawkingrei Both this PR and #5917 are opened when you are a single player in PCP game. I've updated the database to include this 2 PRs and the score into your new team. Check it out at https://pingcap.com/community-cn/tidb-performance-challenge/
